modules/core/include/opencv2/core/base.hpp

```diff
@@ -57,6 +57,172 @@
+#  define CV_DEPRECATED_MIN_DEPTH_DEPTH_ATTR                          CV_DEPRECATED_MSG(CV_DEPRECATED_MIN_DEPTH_DEPTH)
+#  define CV_DEPRECATED_ELEMDEPTH_CMP_ELEMTYPE_ATTR                   CV_DEPRECATED_MSG(CV_DEPRECATED_ELEMDEPTH_CMP_ELEMTYPE)
+#  define CV_DEPRECATED_ELEMDEPTH_TO_ELEMTYPE_ATTR(depr_var, new_var) CV_DEPRECATED_MSG(CV_DEPRECATED_PARAM(ElemDepth, depr_var, ElemType, new_var))
+#  define CV_DEPRECATED_ELEMTYPE_TO_ELEMDEPTH_ATTR(depr_var, new_var) CV_DEPRECATED_MSG(CV_DEPRECATED_PARAM(ElemType, depr_var, ElemDepth, new_var))
+#else
+#  define CV_DEPRECATED_MAX_DEPTH_DEPTH_ATTR                          /* nothing, default */
+#  define CV_DEPRECATED_MIN_DEPTH_DEPTH_ATTR                          /* nothing, default */
+#  define CV_DEPRECATED_ELEMDEPTH_CMP_ELEMTYPE_ATTR                   /* nothing, default */
+#  define CV_DEPRECATED_ELEMDEPTH_TO_ELEMTYPE_ATTR(depr_var, new_var) /* nothing, default */
+#  define CV_DEPRECATED_ELEMTYPE_TO_ELEMDEPTH_ATTR(depr_var, new_var) /* nothing, default */
+#endif
+
+#ifndef OPENCV_DISABLE_DEPRECATED_WARNING_INT_ELEMTYPE_OVERLOAD
+#  define CV_DEPRECATED_MAX_DEPTH_INT_ATTR                            CV_DEPRECATED_MSG(CV_DEPRECATED_MAX_DEPTH_INT)
+#  define CV_DEPRECATED_MIN_DEPTH_INT_ATTR                            CV_DEPRECATED_MSG(CV_DEPRECATED_MIN_DEPTH_INT)
+#  define CV_DEPRECATED_ELEMTYPE_CMP_INT_ATTR                         CV_DEPRECATED_MSG(CV_DEPRECATED_ELEMTYPE_CMP_INT)
+#  define CV_DEPRECATED_ELEMDEPTH_CMP_INT_ATTR                        CV_DEPRECATED_MSG(CV_DEPRECATED_ELEMDEPTH_CMP_INT)
+#  define CV_DEPRECATED_INT_TO_ELEMDEPTH_ATTR(depr_var, new_var)      CV_DEPRECATED_MSG(CV_DEPRECATED_PARAM(int, depr_var, ElemDepth, new_var))
+#  define CV_DEPRECATED_INT_TO_ELEMTYPE_ATTR(depr_var, new_var)       CV_DEPRECATED_MSG(CV_DEPRECATED_PARAM(int, depr_var, ElemType, new_var))
+#else
+#  define CV_DEPRECATED_MAX_DEPTH_INT_ATTR                            /* nothing */
+#  define CV_DEPRECATED_MIN_DEPTH_INT_ATTR                            /* nothing */
+#  define CV_DEPRECATED_ELEMTYPE_CMP_INT_ATTR                         /* nothing */
+#  define CV_DEPRECATED_ELEMDEPTH_CMP_INT_ATTR                        /* nothing */
+#  define CV_DEPRECATED_INT_TO_ELEMDEPTH_ATTR(depr_var, new_var)      /* nothing */
+#  define CV_DEPRECATED_INT_TO_ELEMTYPE_ATTR(depr_var, new_var)       /* nothing */
+#endif
+
+namespace std
+{
+    CV_DEPRECATED_MAX_DEPTH_INT_ATTR static inline int max(const ElemDepth left, const int right)
+    {
+        return max(static_cast<int>(left), right);
+    }
+
+    CV_DEPRECATED_MAX_DEPTH_INT_ATTR static inline int max(const int left, const ElemDepth right)
+    {
+        return max(left, static_cast<int>(right));
+    }
+
+    CV_DEPRECATED_MAX_DEPTH_DEPTH_ATTR static inline int max(const ElemDepth left, const ElemDepth right)
+    {
+        return max(static_cast<int>(left), static_cast<int>(right));
```

Comment:
> Do we really need to deprecate this (Depth, Depth) overload?

If we want to avoid `CV_16F > CV_32F` evaluation as true, then yes we do.

> CV_FP16 brokes existed ordering

We can manage that inside `CV_MAX_DEPTH` macro.

> (Type, Type) should be deprecated

I am not sure I have seen such comparisons, but it makes sense

Reply:
I gave it another thought, and I realized that deprecation is about dropping support, while
> (Type, Type) 

was not supported to begin with. Therefore, allow me to skip this, otherwise, if we try to mark each stupid practice as a depreciated one, we will have thousands of lines to write.
I guess it is indispensable for reviewers to catch such abuses.